File: leetcode/algorithms/cpp/Sort Colors/Sort Colors.cpp
#include <vector>

using namespace std;

class Solution {
public:
  // 88/215

  void sortColors(vector<int> &nums) {
    // nums[0..zero] 0
    // nums[zero+1,two-1] 1
    // nums[two, n-1] 2
    size_t n = nums.size();
    int zero = -1, two = n;
    for (int i = 0; i < two;) {
      if (nums[i] == 1) {
        i++;
      } else if (nums[i] == 2) {
        swap(nums[--two], nums[i]);
      } else { // 0
        assert(nums[i] == 0);
        swap(nums[++zero], nums[i++]);
      }
    }
  }
};

File: leetcode/algorithms/Sort Colors/sort-colors.py
from typing import List


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        lo, cur, hi = 0, 0, len(nums) - 1
        # [lo, hi] 全部都是1
        while cur <= hi:
            if nums[cur] == 0:
                nums[lo], nums[cur] = nums[cur], nums[lo]
                cur += 1
                lo += 1
            elif nums[cur] == 1:
                cur += 1
            else:
                nums[hi], nums[cur] = nums[cur], nums[hi]
                hi -= 1
